librepo/handle.c

```diff
@@ -841,7 +867,57 @@ lr_handle_setopt(LrHandle *handle,
     return ret;
 }
 
-static gboolean
+gboolean
+lr_handle_network_wait(LrHandle *handle, GError **err, guint seconds)
+{
+
+    assert(!err || *err == NULL);
+
+    if (!handle) {
+        g_set_error(err, LR_HANDLE_ERROR, LRE_BADFUNCARG,
+                    "No handle specified");
+        return FALSE;
+    }
+
+    GNetworkMonitor *monitor = g_network_monitor_get_default();
+    GCancellable *cancellable = g_cancellable_new ();
+
+    struct callback_data data_struct;
+    data_struct.input_seconds = seconds;
+    data_struct.begin = time(NULL);
+    data_struct.cancellable = cancellable;
+    data_struct.monitor = monitor;
+
+    gchar *baseurl;
+    if(handle->metalinkurl)
+      baseurl = handle->metalinkurl;
+    else if(handle->mirrorlisturl)
+      baseurl = handle->mirrorlisturl;
+    else if(handle->urls)
+      baseurl = handle->urls[0];
+    assert(baseurl);
+
+    GUri *uri = g_uri_parse(baseurl, G_URI_FLAGS_NONE, NULL);
+    if(uri == NULL){
+      return FALSE;
+    }
+    const gchar* host = g_uri_get_host(uri);
+    guint16 port = g_uri_get_port(uri);
+    GSocketConnectable *connectable =  g_network_address_new(host, port);
+    data_struct.connectable = connectable;
+    GMainLoop *loop;
+    loop = g_main_loop_new(NULL, FALSE);
+    data_struct.loop = loop;
+    g_timeout_add(200, timeout_callback, &data_struct); 
```

Comment:
Hmm.  I'd hoped there was a way with GNetworkMonitor to be event driven, but it seems like there's actually not.  That means that this use of a mainloop is not functionally different from the previous `while` loop you had.  (But, I guess it's slightly easier to read?)

That said, what we *can* do is monitor in an event-driven way the `network-available` boolean.  If that's `FALSE` then we can reasonably assume that no remote hosts are reachable.

But eh...it doesn't matter too much I guess.

Reply:
I am unable to comprehend why is the current implementation using GMainEventLoop similar to the "while" loop implementation. It actively waits on the g_network_monitor_can_reach till the network-timeout expires. Also to be able to monitor the network-available boolean in an event-driven manner, I would still need to call the g_network_monitor_can_reach and set boolean variable to true/false appropriately. So how do the two methods differ?

Reply:
Right, but as I understand it `g_network_monitor_can_reach` will immediately return if the host is unreachable (e.g. if there's no default route).  Then 200 milliseconds later, we'll wake up and try again.

IOW, it's still polling - and that makes it similar to the `while` loop.

> Also to be able to monitor the network-available boolean in an event-driven manner, I would still need to call the g_network_monitor_can_reach and set boolean variable to true/false appropriately. 

Yes, but if I understand things correctly, if the machine is completely offline (e.g. cloud system that is still booting up), then the `network-available` boolean will be `FALSE` and we *know* that connections won't work.

But, very quickly after DHCP works and we get a route, then we'll get a signal from NetworkManager (or kernel netlink) and that will wake the process up.

Another way to look at this is: it's an optimization to avoid polling when there's no network.

(To be very clear: I *could be wrong* about some of the details here!)

Reply:
Ok that makes sense. Now to implement:

> monitor in an event-driven way the network-available boolean

Option 1: I tried exploring creating custom events in the GMainEventLoop using `g_source_new` and using [GSourceFuncs](https://www.freedesktop.org/software/gstreamer-sdk/data/docs/2012.5/glib/glib-The-Main-Event-Loop.html#GSourceFuncs) such as prepare, check, dispatch, finalize to implement the logic. An example of [Creating a new source (scroll down)](https://docs.huihoo.com/symbian/nokia-symbian3-developers-library-v0.8/GUID-7FD05006-09C1-4EF4-A2EB-AD98C2FA8866.html). These custom events are not time based and hence won't poll at a specific interval. The logic of callback function will almost remain the same, but the GSourceFuncs will be added.
I am not sure whether this is still method will still poll like a simple `while` loop implementation.

Option2: Is there any other method not involving GMainEventLoop at all?

Reply:
Use `g_signal_connect("notify::network-available")` on the monitor instance, pass the main loop as user data.  When the `network-available` property becomes `TRUE`, you could exit the loop, and then add the timer to the loop and restart it.